pkg/webhook/preflight/preflight.go

```diff
@@ -0,0 +1,153 @@
+	Init(ctx context.Context, client ctrlclient.Client, cluster *clusterv1.Cluster) []Check
+}
+
+type WebhookHandler struct {
+	client   ctrlclient.Client
+	decoder  admission.Decoder
+	checkers []Checker
+}
+
+func New(client ctrlclient.Client, decoder admission.Decoder, checkers ...Checker) *WebhookHandler {
+	h := &WebhookHandler{
+		client:   client,
+		decoder:  decoder,
+		checkers: checkers,
+	}
+	return h
+}
+
+func (h *WebhookHandler) Handle(ctx context.Context, req admission.Request) admission.Response {
+	if req.Operation == admissionv1.Delete {
+		return admission.Allowed("")
+	}
+
+	cluster := &clusterv1.Cluster{}
+	err := h.decoder.Decode(req, cluster)
+	if err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	// Checks run only for ClusterClass-based clusters.
+	if cluster.Spec.Topology == nil {
+		return admission.Allowed("")
+	}
+
+	resp := admission.Response{
+		AdmissionResponse: admissionv1.AdmissionResponse{
+			Allowed: true,
+			Result: &metav1.Status{
+				Details: &metav1.StatusDetails{},
+			},
+		},
+	}
+
+	// Collect all checks in parallel.
+	checkerWG := &sync.WaitGroup{}
+	resultCh := make(chan CheckResult)
+	for _, checker := range h.checkers {
+		checkerWG.Add(1)
+
+		go func(ctx context.Context, checker Checker, resultCh chan CheckResult) {
+			// Initialize the checker.
+			checks := checker.Init(ctx, h.client, cluster)
+
+			// Run its checks in parallel.
+			checksWG := &sync.WaitGroup{}
+			for _, check := range checks {
+				checksWG.Add(1)
+				go func(ctx context.Context, check Check, resultCh chan CheckResult) {
+					result := check(ctx)
+					resultCh <- result
+					checksWG.Done()
+				}(ctx, check, resultCh)
+			}
+			checksWG.Wait()
+
+			checkerWG.Done()
+		}(ctx, checker, resultCh)
+	}
+
+	// Close the channel when all checkers are done.
+	go func(wg *sync.WaitGroup, resultCh chan CheckResult) {
+		wg.Wait()
+		close(resultCh)
+	}(checkerWG, resultCh)
+
+	// Collect all check results from the channel.
+	internalError := false
+	for result := range resultCh {
+		if result.Error {
+			internalError = true
+		}
+
+		if !result.Allowed {
+			resp.Allowed = false
+		}
+
+		for _, cause := range result.Causes {
+			resp.Result.Details.Causes = append(resp.Result.Details.Causes, metav1.StatusCause{
+				Type:    metav1.CauseType(fmt.Sprintf("FailedPreflight%s", result.Name)),
+				Message: cause.Message,
+				Field:   cause.Field,
+			})
+		}
+
+		resp.Warnings = append(resp.Warnings, result.Warnings...)
+	}
+
+	if len(resp.Result.Details.Causes) == 0 {
+		return resp
+	}
+
+	// Because we have some causes, we construct the response message and code.
+	resp.Result.Message = "preflight checks failed"
+	resp.Result.Code = http.StatusForbidden
+	resp.Result.Reason = metav1.StatusReasonForbidden
+	if internalError {
+		// Internal errors take precedence over check failures.
+		resp.Result.Code = http.StatusInternalServerError
+		resp.Result.Reason = metav1.StatusReasonInternalError
+	}
+
```

Comment:
I don't checking if causes is empty is enough here, does it handle the error response correctly? How about a switch instead to cover error and allowed state? Also think it should use `StatusReasonInvalid` rather than `StatusReasonForbidden`.

```suggestion
	switch {
		case internalError:
			// Internal errors take precedence over check failures.
			resp.Result.Code = http.StatusInternalServerError
			resp.Result.Reason = metav1.StatusReasonInternalError
		case !resp.Allowed:
			// Because the response is not allowed, preflights must have failed.
			resp.Result.Message = "preflight checks failed"
			resp.Result.Code = http.StatusInvalid
			resp.Result.Reason = metav1.StatusReasonInvalid
	}

```